module_utils/network_lsr/nm/client.py

```diff
@@ -0,0 +1,84 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+import logging
+
+# Relative import is not support by ansible 2.8 yet
+# pylint: disable=import-error, no-name-in-module
+from ansible.module_utils.network_lsr.nm.error import LsrNetworkNmError  # noqa:E501
+
+import gi
+
+gi.require_version("NM", "1.0")
+
+# It is required to state the NM version before importing it
+# But this break the flake8 rule: https://www.flake8rules.com/rules/E402.html
+# Use NOQA: E402 to suppress it.
+from gi.repository import NM  # NOQA: E402
+from gi.repository import GLib  # NOQA: E402
+from gi.repository import Gio  # NOQA: E402
+
+# pylint: enable=import-error, no-name-in-module
+
+NM
+GLib
+Gio
+
+
+def get_client():
+    return NM.Client.new()
+
+
+class _NmMainLoop(object):
+    def __init__(self, timeout):
+        self._mainloop = GLib.MainLoop()
+        self._cancellable = Gio.Cancellable.new()
+        self._timeout = timeout
+        self._timeout_id = None
+
+    def run(self):
+        logging.debug("NM mainloop running")
+        user_data = None
+        self._timeout_id = GLib.timeout_add(
+            int(self._timeout * 1000),
+            self._timeout_call_back,
+            user_data,
+        )
+        logging.debug("Added timeout checker")
+        self._mainloop.run()
+
+    def _timeout_call_back(self, _user_data):
+        logging.error("Timeout")
+        self.fail(LsrNetworkNmError("Timeout"))
+
+    @property
+    def cancellable(self):
+        return self._cancellable
+
+    @property
+    def is_cancelled(self):
+        return self._cancellable.is_cancelled()
+
+    def _clean_up(self):
+        logging.debug("NM mainloop cleaning up")
+        if self._timeout_id:
+            logging.debug("Removing timeout checker")
+            GLib.source_remove(self._timeout_id)
+            self._timeout_id = None
+        if self._cancellable:
+            logging.debug("Canceling all pending tasks")
+            self._cancellable.cancel()
+            self._cancellable = None
+        self._mainloop = None
+
+    def quit(self):
+        logging.debug("NM mainloop quiting")
+        self._mainloop.quit()
+        self._clean_up()
+
+    def fail(self, exception):
+        self.quit()
+        raise exception
```

Comment:
I am surprised this works. `fail` is commonly called while running the mainloop, so how does the python exception get propagated through the C code (in glib, which runs the mainloop)? According to https://stackoverflow.com/questions/54666827/pygobject-glib-mainloop-and-exceptions, it would seem that this does not work. But obviously, it does somehow.

Reply:
The `raise` is ran after mainloop quit. So the exception is not handlled in side mainloop.

Reply:
that's not what `g_main_loop_quit()` does. `GMainLoop` is basically a boolean flag (`g_main_loop_is_running()`) and two functions:

  - `g_main_loop_run()` sets is-running flag to true, and does `while (is_running) iterate-main-context` ([[1]](https://gitlab.gnome.org/GNOME/glib/-/blob/b8927cc6ad866570b100aa71e5aeabd349f439f3/glib/gmain.c#L4316)).
  - `g_main_loop_quit()` sets the is-flag to false. Nothing else ([[2]](https://gitlab.gnome.org/GNOME/glib/-/blob/b8927cc6ad866570b100aa71e5aeabd349f439f3/glib/gmain.c#L4343)).

`g_main_loop_quit()`  cannot jump to the call-stack of where `g_main_loop_run()` is invoked. It clears the `is_running` flag. Consequently, the calls after `g_main_loop_quit()` call still are invoked in the same frame of the call stack.

Reply:
> So the exception is not handlled in side mainloop.

Hm, guess it depends on what you mean with "in side mainloop". I would say that code executes "inside the mainloop" if the current call-stack has `g_main_context_iterate()` in the parent frames. That means, the current code is invoked as part of a callback function to a `GSource` which the `GMainContext` is currently dispatching. With that definition, the `raise` certainly happens inside the mainloop.

there is also a warning in the docs for `g_main_loop_quit()`:
```
 * Note that sources that have already been dispatched when 
 * g_main_loop_quit() is called will still be executed.
```

that means, after a call to quit(), not only does the current call-stack come to it's normal end (there is no jump/exception), even more sources might be dispatched (if they were already ready) before `g_main_loop_run()` returns.

Reply:
If you have suggestions on how to raise exception found in mainllop, please state a demo code.
If you have questions regarding why this code works, I would suggest reach out experts(not me) somewhere else.

For me, the code works as expected. And I cannot switch to `NM.Client.get_main_context()` as RHEL 7 still hold very old NetworkManager.